### PR TITLE
Lollipop plot in react

### PR DESCRIPTION
**Pull request by adamabeshouse**
# Checks
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Airbnb React/JSX Style guide](https://github.com/airbnb/javascript/tree/master/react).
- [ ] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)

### CODE REVIEW

File: src/pages/patientView/PatientViewPage.tsx [hunk 1]
```diff
@@ -3,7 +3,7 @@ import * as _ from 'lodash';
 import $ from 'jquery';
 import { default as ReactBootstrap} from 'react-bootstrap';
 import GenomicOverview from './genomicOverview/GenomicOverview';
-import { ClinicalData } from "shared/api/generated/CBioPortalAPI";
+import {ClinicalData, Mutation} from "shared/api/generated/CBioPortalAPI";
```
Comment (adamabeshouse) [inline src/pages/patientView/PatientViewPage.tsx]: unused import

File: src/pages/patientView/mutation/column/AlleleFreqColumnFormatter.tsx [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import * as React from 'react';
 import {If, Else, Then } from 'react-if';
-import DefaultTooltip from "shared/components/DefaultTooltip";
+import DefaultTooltip from "../../../../shared/components/DefaultTooltip";
```
Comment (adamabeshouse) [inline src/pages/patientView/mutation/column/AlleleFreqColumnFormatter.tsx]: unnecessary change

File: src/pages/patientView/sampleManager.tsx [hunk 3]
```diff
@@ -3,7 +3,7 @@ import * as _ from 'lodash';
 import SampleInline from './patientHeader/SampleInline';
 import {ClinicalDataBySampleId} from "../../shared/api/api-types-extended";
 import ClinicalInformationPatientTable from "./clinicalInformation/ClinicalInformationPatientTable";
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../../shared/components/DefaultTooltip';
```
Comment (adamabeshouse) [inline src/pages/patientView/sampleManager.tsx]: unnecessary change

File: src/pages/patientView/vafPlot/ThumbnailExpandVAFPlot.tsx [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../../../shared/components/DefaultTooltip';
```
Comment (adamabeshouse) [inline src/pages/patientView/vafPlot/ThumbnailExpandVAFPlot.tsx]: unnecessary change

File: src/shared/components/MouseEventInterceptor.tsx [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import * as React from "react";
```
Comment (adamabeshouse) [inline src/shared/components/MouseEventInterceptor.tsx]: this file can be deleted

File: src/shared/components/annotation/CancerHotspots.tsx [hunk 6]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/annotation/CancerHotspots.tsx]: other way is better

File: src/shared/components/annotation/MyCancerGenome.tsx [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import * as React from 'react';
 import * as _ from 'lodash';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/annotation/MyCancerGenome.tsx]: other way is better

File: src/shared/components/annotation/OncoKB.tsx [hunk 8]
```diff
@@ -2,7 +2,7 @@ import * as React from 'react';
 import {Modal} from 'react-bootstrap';
 import {observer} from "mobx-react";
 import {Circle} from "better-react-spinkit";
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/annotation/OncoKB.tsx]: other way is better

File: src/shared/components/annotation/OncoKbCard.tsx [hunk 9]
```diff
@@ -2,7 +2,7 @@ import * as React from 'react';
 import * as _ from 'lodash';
 import Collapse from 'react-collapse';
 import {If, Then, Else} from 'react-if';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/annotation/OncoKbCard.tsx]: other way is better

File: src/shared/components/cohort/FrequencyBar.tsx [hunk 10]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/cohort/FrequencyBar.tsx]: other way is better

File: src/shared/components/copyDownloadControls/CopyDownloadControls.tsx [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import * as React from 'react';
 import {Button, ButtonGroup} from 'react-bootstrap';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/copyDownloadControls/CopyDownloadControls.tsx]: other way is better

File: src/shared/components/lollipopMutationPlot/Lollipop.tsx [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    stickColor?:string;
+    label?:string;
+    hitzoneClassName?:string;
+};
+
+@observer
+export default class Lollipop extends React.Component<LollipopProps, {}> {
+    private handlers:any;
+
+    private noop:()=>void = ()=>{};
```
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/Lollipop.tsx]: unused

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlot.tsx [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    lollipops:LollipopSpec[];
+    domains:DomainSpec[];
+    vizWidth:number;
+    vizHeight:number;
+    geneCodonLength:number;
+    yMax?:number;
+    onSelectionChange?:(selectedPositions:{ [pos:number]:boolean} )=>void;
+};
+
+const DELETE_FOR_DOWNLOAD_CLASS = "delete-for-download";
```
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlot.tsx]: unused

File: src/shared/components/lollipopMutationPlot/styles.scss [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  visibility:hidden;
+}
+.fade-in {
+  -webkit-transition: opacity 0.2s;
+  -moz-transition: opacity 0.2s;
+  -o-transition: opacity 0.2s;
+  transition: opacity 0.2s;
+  opacity: 1;
+}
+.collapse-out {
```
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/styles.scss]: unused because not currently working.
@onursumer do you think its necessary to keep the height collapsing animation?

File: src/shared/components/mutationTable/column/CosmicColumnFormatter.tsx [hunk 15]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/mutationTable/column/CosmicColumnFormatter.tsx]: other way is better

File: src/shared/components/mutationTable/column/DiscreteCNAColumnFormatter.tsx [hunk 16]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/mutationTable/column/DiscreteCNAColumnFormatter.tsx]: other way is better

File: src/shared/components/mutationTable/column/MrnaExprColumnFormatter.tsx [hunk 17]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/mutationTable/column/MrnaExprColumnFormatter.tsx]: other way is better

File: src/shared/components/mutationTable/column/MutationAssessorColumnFormatter.tsx [hunk 18]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/mutationTable/column/MutationAssessorColumnFormatter.tsx]: other way is better

File: src/shared/components/mutationTable/column/MutationTypeColumnFormatter.tsx [hunk 19]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/mutationTable/column/MutationTypeColumnFormatter.tsx]: other way is better

File: src/shared/components/mutationTable/column/SampleColumnFormatter.tsx [hunk 20]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/mutationTable/column/SampleColumnFormatter.tsx]: other way is better

File: src/shared/components/mutationTable/column/TumorAlleleFreqColumnFormatter.tsx [hunk 21]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import DefaultTooltip from "shared/components/DefaultTooltip";
+import DefaultTooltip from "../../DefaultTooltip";
```
Comment (adamabeshouse) [inline src/shared/components/mutationTable/column/TumorAlleleFreqColumnFormatter.tsx]: other way is better

File: src/shared/components/tableHeaderControls/TableHeaderControls.tsx [hunk 22]
```diff
@@ -1,6 +1,6 @@
 import * as React  from 'react';
 import * as ReactDOM from 'react-dom';
-import DefaultTooltip from 'shared/components/DefaultTooltip';
+import DefaultTooltip from '../DefaultTooltip';
```
Comment (adamabeshouse) [inline src/shared/components/tableHeaderControls/TableHeaderControls.tsx]: other way is better

**Comment by adamabeshouse** [comment]
[Results Page: Mutations Tab: Lollipop Diagram (8 remaining)](https://trello.com/c/AUONzplj/226-results-page-mutations-tab-lollipop-diagram-8-remaining)

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx [hunk 23]
```diff
@@ -0,0 +1,508 @@
...
+    }
+}
+
+const MISSENSE_COLOR = "#008000";
+const TRUNCATING_COLOR = "#000000";
+const INFRAME_COLOR = "#8B4513";
+const OTHER_COLOR = "#8B00C9";
+
+@observer
+export default class LollipopMutationPlotWrapper extends React.Component<LollipopMutationPlotWrapperProps, {}> {
```
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: Hmm, how about `MutationDiagram`? :)

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx [hunk 24]
```diff
@@ -0,0 +1,508 @@
...
+                    return INFRAME_COLOR;
+                default:
+                    return OTHER_COLOR;
+            }
+        } else {
+            return "#ff0000"; // we only get here if theres no mutations, which shouldnt happen. red to indicate an error
+        }
+    }
+
+    readonly swissProtId = remoteData({
```
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: It might be better to move this, and other `remoteData` functions into the store or into separate cache instances

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx [hunk 25]
```diff
@@ -0,0 +1,508 @@
...
+                        {' '}
+                        <FormGroup>
+                            <Button onClick={this.handlers.handlePDFClick}>
+                                PDF
+                            </Button>
+                        </FormGroup>
+                        {'  '}
+                        <FormGroup>
+                            <InputGroup>
+                                <InputGroup.Addon>Y-Axis Max:</InputGroup.Addon>
```
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: Maybe we should also have the y-axis min and set it to 5 by default (legacy setting).

File: src/shared/lib/StringUtils.ts [hunk 26]
```diff
@@ -0,0 +1,18 @@
+export function longestCommonStartingSubstring(str1:string, str2:string) {
+    // from https://github.com/cBioPortal/mutation-mapper/blob/master/src/js/util/cbio-util.js
```
Comment (onursumer) [inline src/shared/lib/StringUtils.ts]: Looks very familiar ;)

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotNoTooltip.tsx [hunk 27]
```diff
@@ -0,0 +1,349 @@
...
+
+    private codonToX(codon:number) {
+        return (codon/this.props.geneCodonLength)*this.props.vizWidth;
+    }
+
+    private countToHeight(count:number) {
+        return this.lollipopZeroHeight + Math.min(1, (count/this.yMax))*this.yAxisHeight;
+    }
+    @computed private get lollipopZeroHeight() {
+        return 10;
```
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotNoTooltip.tsx]: Just curious, does computed have any effect on constant values?

File: src/shared/components/lollipopMutationPlot/styles.scss (hunk 14 above)
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/styles.scss]: I think this one is for the PDB rectangles. I don't remember any collapsing animation for the diagram. There was only fading animation for filtering.

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotNoTooltip.tsx (hunk 27 above)
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotNoTooltip.tsx]: good point, probably not

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 25 above)
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: Oh I didn't see that feature in the legacy diagram

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 24 above)
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: we could definitely do that

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 23 above)
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: no Lollipop?

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 24 above)
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: you are right, this is more modular than putting into the store. But, maybe later we add separate cache for each data source in case we need to share the data.

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 23 above)
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: `LollipopMutationDiagram` or `MutationLollipopDiagram`? :)

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 24 above)
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: yeah. i'd say lets cross that bridge when we get there. right now it's not clear to me that every user should be expected to load pfam data

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 23 above)
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: well, the thing is this is the wrapper like, `LollipopMutationPlot` is the view, and this is something that generates the input to the view. so Im not sure if diagram captures that

File: src/shared/components/lollipopMutationPlot/styles.scss (hunk 14 above)
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/styles.scss]: the collapsing was just for the color code panel

Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/styles.scss]: Oh, that one. I think we can live without the animation. And if we really need the animation, there is this library `react-collapse`, which is already included in the project.

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 25 above)
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: We don't need to stick to that default, but it looks better to me when there are few mutations. Maybe I am too much used to it...

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 23 above)
Comment (onursumer) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: Oh, you are right. Actually, how about renaming the low level plot as just `LollipopPlot` and the wrapper as `LollipopMutationPlot`?

Ideally with `LollipopPlot` you can visualize any set of data grouped by codon/location. And maybe rename `Domains` to something more generic, like `XAxisOverlay` (or something nicer).

Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: for now maybe Domain is okay? but I renamed geneCodonLength to xMax

File: src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx (hunk 25 above)
Comment (adamabeshouse) [inline src/shared/components/lollipopMutationPlot/LollipopMutationPlotWrapper.tsx]: @jjgao what do you think?

**Comment by adamabeshouse** [comment]
[Mutation type count panel](https://trello.com/c/dt2nEd2G/272-mutation-type-count-panel)

File: src/shared/components/lollipopMutationPlot/LollipopPlotNoTooltip.tsx [hunk 28]
```diff
@@ -0,0 +1,335 @@
...
+                }
+                this.selectedPositions.set(codon+"", !prevSelection);
+                this.props.onSelectionChange && this.props.onSelectionChange(this.selectedPositions.toJS());
+            }
+        };
+    }
+
+
+    componentDidMount() {
+        $(document).keydown(e=>{
```
Comment (alisman) [inline src/shared/components/lollipopMutationPlot/LollipopPlotNoTooltip.tsx]: can we get a comment on this behavior?

Comment (alisman) [inline src/shared/components/lollipopMutationPlot/LollipopPlotNoTooltip.tsx]: you should kill these listeners on unmount of component. otherwise, i believe there will be a memory leak.

File: src/shared/components/mutationTypePanel/LollipopMutationTypePanel.tsx [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+    missenseColor:string;
+    inframeColor:string;
+    truncatingColor:string;
+    otherColor:string;
+};
+
+const buttonOrder:LollipopMutationType[] = ["missense", "truncating", "inframe", "other"];
+
+@observer
+export default class LollipopMutationTypePanel extends React.Component<LollipopMutationTypePanelProps, {}> {
```
Comment (onursumer) [inline src/shared/components/mutationTypePanel/LollipopMutationTypePanel.tsx]: `Lollipop` does not feel right here. Because there are no lollipops in this panel.

Can we name this one to `MutationTypePanel` and the lower level component to something generic like `TypePanel`, `CountPanel`, `TypeCountPanel`, or something better?

File: src/shared/lib/getCanonicalMutationType.ts [hunk 30]
```diff
@@ -1,9 +1,28 @@
...
+
 export type CanonicalMutationType =
-    "missense" | "frame_shift_ins" | "frame_shift_del" | "frameshift" |
-        "nonsense" | "splice_site" | "nonstart" | "nonstop" | "in_frame_del" |
-        "in_frame_ins" | "inframe" | "truncating" | "fusion" | "silent" | "other";
+    MISSENSE | FRAME_SHIFT_INS | FRAME_SHIFT_DEL | FRAMESHIFT |
+        NONSENSE | SPLICE_SITE | NONSTART | NONSTOP | IN_FRAME_DEL |
+        IN_FRAME_INS | INFRAME | TRUNCATING | FUSION | SILENT | OTHER;
+
+export type LollipopMutationType =
```
Comment (onursumer) [inline src/shared/lib/getCanonicalMutationType.ts]: I think this mutation type should have a different name. `Lollipop` is a visualization term, but these fewer types we choose are something conceptually separate from the lollipop. I don't really know what name fits best here. In legacy implementation, I called them `main` types, but that is not a good name either.

**Comment by adamabeshouse** [comment]
can mouseenter events on lollipops and domains be assigned using event delegation? so that you only set one listener?

File: src/shared/components/lazyMobXTable/LazyMobXTable.tsx [hunk 31]
```diff
@@ -391,25 +349,35 @@ class LazyMobXTableStore<T> {
...
+                return (<td key={column.name}>
+                    {column.render(datum)}
+                </td>);
+            });
+            return _tds;
         });
     }
 
+    @computed get rows():JSX.Element[] {
+        let x:JSX.Element = (<span/>);
```
Comment (alisman) [inline src/shared/components/lazyMobXTable/LazyMobXTable.tsx]: is x used anywhere?

File: src/shared/components/lazyMobXTable/LazyMobXTable.tsx [hunk 32]
```diff
@@ -544,58 +512,66 @@ export default class LazyMobXTable<T> extends React.Component<LazyMobXTableProps
...
+                onNextPageClick={this.handlers.incPage}
+                textBetweenButtons={this.store.paginationStatusText}
+                previousPageDisabled={this.store.page === 0}
+                nextPageDisabled={this.store.page === this.store.maxPage}
+                {...this.props.paginationProps}
+            />
+        );
+    }
+
+    @computed get toolbar() {
```
Comment (alisman) [inline src/shared/components/lazyMobXTable/LazyMobXTable.tsx]: we should discuss using computed on jsx like this.  the reason it worries me is that it makes the assumption that every dependency referenced is observable.  It replaces Mobx's practice of tagging components as Observers, which has the advantage that it implements shouldComponentUpdate, which can do a shallow comparison on dependencies which happen not to be observable.   Using computed to avoid React render cycles is another paradigm and we should make sure we really want to do this.

Comment (adamabeshouse) [inline src/shared/components/lazyMobXTable/LazyMobXTable.tsx]: I see, you mean as opposed to moving this stuff out into a separate Observer component. That sounds like a good idea to me

File: src/shared/components/lazyMobXTable/LazyMobXTable.tsx (hunk 31 above)
Comment (adamabeshouse) [inline src/shared/components/lazyMobXTable/LazyMobXTable.tsx]: no, oversight from testing

File: src/shared/components/lazyMobXTable/LazyMobXTable.tsx (hunk 32 above)
Comment (adamabeshouse) [inline src/shared/components/lazyMobXTable/LazyMobXTable.tsx]: after conv I think it makes sense to keep this in a @computed, because
(1) mobx @observer makes props observable
(2) anything that mobx wont catch this way, react also wouldnt be expected to catch (i.e. a deep prop change)

**Comment by adamabeshouse** [comment]
bug: 
image
